aslam_cv_pipeline/src/visual-npipeline.cc

```diff
@@ -76,37 +78,44 @@ bool VisualNPipeline::processImageNonBlockingDroppingOldestNFrameIfFull(
     size_t max_output_queue_size) {
   CHECK_GE(max_output_queue_size, 1u);
 
-  bool result = false;
   std::unique_lock<std::mutex> lock(mutex_);
+  if (shutdown_) {
+    return false;
+  }
   if (completed_.size() >= max_output_queue_size) {
     completed_.erase(completed_.begin());
-    result = true;
+    condition_not_full_.notify_all();
   }
-  processImage(camera_index, image, timestamp);
-  return result;
+  processImageImpl(camera_index, image, timestamp);
+  return !shutdown_;
 }
 
 bool VisualNPipeline::getNextBlocking(std::shared_ptr<VisualNFrame>* nframe) {
   CHECK_NOTNULL(nframe);
+
   std::unique_lock<std::mutex> lock(mutex_);
-  if (shutdown_) {
-    return false;
-  }
-  if (completed_.empty()) {
-    condition_not_empty_.wait(lock);
+  while (!shutdown_) {
     if (completed_.empty()) {
-      // Queue shutdown in the meantime.
-      return false;
+      condition_not_empty_.wait(lock);
+      if (shutdown_) {
+        return false;
+      }
+      if (completed_.empty()) {
+        continue;
+      }
     }
+    // Get the oldest frame.
+    *nframe = getNextImpl();
+    CHECK(*nframe);
+    return !shutdown_;
```

Comment:
hm I guess the question is whether you still want to have the "last" valid nframe processed before the pipeline shut down. I mean I don't really see why you would not want to have this last valid nframe processed upstream. Wdyt?

Reply:
I was thinking of the same. However, then the return value would not reflect the shutdown-state of the pipeline. The same question arises for the [thread-safe-queue](https://github.com/ethz-asl/maplab/blob/master/common/maplab-common/include/maplab-common/threadsafe-queue.h). At least we would have to adapt the description to: Returns true if the nframe retrieval was successful. Otherwise, returns false which implies that the queue has shut down.

Reply:
adapting the description makes a tiny bit more sense to me (because again, I don't see why you would not want to have this nframe processed), but all in all, it's a close call and I'm fine either way. can I leave it up to you to decide?

Reply:
decision made: will adapt description.